x-pack/plugins/security_solution/public/one_discover/cell_renderers/cell_renderers.tsx

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import React, { useMemo } from 'react';
+import type { DataGridCellValueElementProps } from '@kbn/unified-data-table';
+import type { TimelineNonEcsData } from '@kbn/timelines-plugin/common';
+import type { SecuritySolutionCellRenderFeature } from '@kbn/discover-shared-plugin/public';
+import type { ColumnHeaderType } from '../../../common/types';
+import type { Maybe } from '../../../common/search_strategy';
+import { DefaultCellRenderer } from '../../timelines/components/timeline/cell_rendering/default_cell_renderer';
+import { ONE_DISCOVER_SCOPE_ID } from '../constants';
+
+export type SecuritySolutionRowCellRendererGetter = Awaited<
+  ReturnType<SecuritySolutionCellRenderFeature['getRender']>
+>;
+
+const ALLOWED_DISCOVER_RENDERED_FIELDS = ['host.name', 'user.name', 'source.ip', 'destination.ip'];
+
+export const getCellRendererForGivenRecord: SecuritySolutionRowCellRendererGetter = (
+  fieldName: string
+) => {
+  if (!ALLOWED_DISCOVER_RENDERED_FIELDS.includes(fieldName)) return undefined;
+  return function UnifiedFieldRenderBySecuritySolution(props: DataGridCellValueElementProps) {
+    // convert discover data format to timeline data format
+    const data: TimelineNonEcsData[] = useMemo(
+      () =>
+        Object.keys(props.row.flattened).map((field) => ({
+          field,
+          value: Array.isArray(props.row.flattened[field])
+            ? (props.row.flattened[field] as Maybe<string[]>)
+            : ([props.row.flattened[field]] as Maybe<string[]>),
+        })),
+      [props.row.flattened]
+    );
```

Comment:
It's not as important while the profile is experimental, but I think we'll run into performance issues iterating over all doc fields in every cell renderer like this, especially since some docs can contain hundreds to thousands of fields. I'd recommend finding a more efficient way to handle this longer term.

Reply:
I agree 100%. problem is that data is passed `per row` so as of now I do not see any other way. This will be an issue in other extension points  as well such as `row actions`. 

We are planning to formalize one data format in security solution and we could adapt `Discover's`. I will try to start the discussions soon on that.

Reply:
One alternative could be that I can add `security-like` data format in `row` object below since we are anyway iterating. It could be lazy object so that only when security uses it, it gets executed.  But I undestand in case this is not acceptable as it pollutes discover with unwanted data format.


```ts   
const record: DataTableRecord = { 
     id: String(idx), 
     raw: row, 
     flattened: row, 
     get security_data(){
      ...
      }
   }; 
```
https://github.com/elastic/kibana/blob/7fa9e84ffcb420620bd159669063f9e205aba995/src/plugins/discover/public/application/main/data_fetching/fetch_esql.ts#L87-L95

Reply:
Ideally we wouldn't modify core Discover code to support this, although I think it could technically be achieved with the context awareness framework today. You would need to create a `SecurityDocumentProfileProvider` which matches when in the Security solution view, then format the data within the `resolve` function and return it in a custom context object. Something like this:
```ts
export const createSecurityDocumentProfileProvider = (): DocumentProfileProvider<{
  securityData: TimelineNonEcsData[];
}> => ({
  profileId: 'security-document-profile',
  profile: {},
  resolve: ({ rootContext, record }) => {
    if (rootContext.solutionType !== SolutionType.Security) {
      return { isMatch: false };
    }

    const securityData: TimelineNonEcsData[] = Object.keys(record.flattened).map((field) => ({
      field,
      value: Array.isArray(record.flattened[field])
        ? (record.flattened[field] as Maybe<string[]>)
        : ([record.flattened[field]] as Maybe<string[]>),
    }));

    return {
      isMatch: true,
      context: {
        type: DocumentType.Default,
        securityData,
      },
    };
  },
});
```

And since we store the `context` object on the `record`, you could then access it within your cell renderer using `props.row.context.securityData`. There'd be some casting involved to get the types to work, but otherwise it should be fine.

I wouldn't recommend doing this long term though since it's still not performant or memory efficient compared to working with the `record` directly. Even without Security adopting a unified format, maybe it's possible to decompose the renderers a bit and have separate wrappers for both the Security and Discover formats? Maybe that's easier said than done 😅

Reply:
>And since we store the context object on the record, you could then access it within your cell renderer using props.row.context.securityData. There'd be some casting involved to get the types to work, but otherwise it should be fine.

I did not know that record was available in Document Profiles, this is indeed helpful. But I will most probably change Security itself.. Because currently, we are also transforming data for the `unifed data table` which we would like to stop. Only solution seems that we need to natively support `raw` and `flattened` formats.



>Even without Security adopting a unified format, maybe it's possible to decompose the renderers a bit and have separate wrappers for both the Security and Discover formats

Yes this could be one practical solution. Let me kick off the dicussion and see if it is good time to retire the data format we do have. It is not that efficient anyways being an array.

Reply:
+1 to retiring the data format, but that'll take a good amount of time to accomplish 😓